ddtrace/internal/telemetry/writer.py

```diff
@@ -627,6 +646,29 @@ def _app_dependencies_loaded_event(self, newly_imported_deps: List[str]):
             payload = {"dependencies": packages}
             self.add_event(payload, "app-dependencies-loaded")
 
+    def product_activated(self, product, enabled):
+        # type: (TELEMETRY_APM_PRODUCT, bool) -> None
+        """Adds a Telemetry event which reports the enablement of an APM product"""
+        if self._product_enablement[product.value] == enabled:
+            return
+
+        self._product_enablement[product.value] = enabled
+
+        # the product status will be sent as part of the app_started event
+        if not self.started:
+            return
+
+        payload = {
+            "products": {
+                product.value: {
+                    "version": config.version,
+                    "enabled": enabled,
+                }
+            }
+        }
+
+        self.add_event(payload, "app-product-change")
```

Comment:
We can run into a scenario where a product is enabled, disabled, and then enabled again. This will generate 3 app-product-change payloads. Can we generate the app-product-change payload in `TelemetryWriter.periodic(..)` (similar to how we handle integrations). This way we can batch multiple products into one payload.

```suggestion
```

Reply:
Yes, that makes sense. I think this scenario is unlikely, but then again we cannot assume the logic flow of our users. 

I'll refactor and add an invalidation flag.